upgrade/v17xto180/upgrade.go

```diff
@@ -17,9 +21,176 @@ const (
 )
 
 func UpgradeResources(namespace string, lhClient *lhclientset.Clientset, kubeClient *clientset.Clientset, resourceMaps map[string]interface{}) error {
+	if resourceMaps == nil {
+		return errors.New("resourceMaps cannot be nil")
+	}
+
+	if err := upgradeVolumes(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+
+	if err := upgradeBackupVolumes(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+	if err := upgradeBackups(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+
+	if err := upgradeBackupBackingImages(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+
+	if err := upgradeBackingImageDataSources(namespace, lhClient, resourceMaps); err != nil {
+		return err
+	}
+
 	return upgradeBackingImages(namespace, lhClient, resourceMaps)
 }
 
+func upgradeVolumes(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade volume failed")
+	}()
+
+	volumesMap, err := upgradeutil.ListAndUpdateVolumesInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn volumes during the volume upgrade")
+	}
+
+	for _, v := range volumesMap {
+		backupTargetName := v.Spec.BackupTargetName
+		v.Spec.BackupTargetName = ensureBackupTargetSet(backupTargetName)
+	}
+
+	return nil
+}
+
+func ensureBackupTargetSet(backupTargetName string) string {
+	if backupTargetName == "" {
+		return types.DefaultBackupTargetName
+	}
+	return backupTargetName
+}
+
+func upgradeBackupVolumes(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade backup volume failed")
+	}()
+
+	backupVolumeMap, err := upgradeutil.ListAndUpdateBackupVolumesInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn BackupVolumes during the Longhorn BackupVolume upgrade")
+	}
+
+	for _, bv := range backupVolumeMap {
+		backupTargetName := bv.Spec.BackupTargetName
+		bv.Spec.BackupTargetName = ensureBackupTargetSet(backupTargetName)
+
+		if bv.Spec.VolumeName == "" {
+			bv.Spec.VolumeName = bv.Name
+		}
```

Comment:
What's purpose of the VolumeName?

Reply:
It's used to synchronize the backup volumes between the cluster and the remote backup target.

Reply:
leave a comment for it

Reply:
Added. Thanks